### PR TITLE
Add pmatch command to CLI

### LINKED ISSUE #50: Add ability to efficiently find duplicates in a large dataset to the CLI
Currently, if you have sets of templates (the left and right thumbs for example) from individuals and want to find duplicates within that, you'd need to write a script that compares pairs of templates using `match`. Without a lot of work to deal with this in parallel, this would be very slow as every set has to be compared to each other set.

Keppel's CLI could help with this by providing a command to find matches within a set of data and execute the comparisons in parallel to speed things up. Something like:

```bash
keppel pmatch -i input.csv -o output.csv -t 40 -p 16
```

The input CSV could have columns `id`, `template_1`, `template_2` etc and the output CSV would have `id_1`, `id_2`, `score_1`, `score_2` etc. The -t and -p arguments would allow customising the matching threshold (limiting which pairs get output) and the number of threads used to execute comparisons in parallel respectively.

### PR DESCRIPTION
**Pull request by seadowg**
Closes #50

### CODE REVIEW

**Comment by seadowg** [comment]
@tobiasmcnulty this isn't quite finished yet: I haven't had a chance to try it for real, and there's still some work to spruce up error states and help. I'd also guess there are some more tweaks we can make to extract a little more performance (either real time speed or memory footprint). That all said, I thought I'd ping you that this is in a state where it can be tested in case you were interested in giving it a spin. Currently, you should be able to run it like:

```
keppel pmatch -i input.csv -o output.csv -t 40 -p 16
```

This would check for matches (based on a threshold of 40) between subjects in `input.csv` and write them out to `output.csv` and spin up 16 threads to execute the checks in parallel. The input CSV should be structured like:

```csv
id, template_1, template_2
<id>, <first template hex encoded>, <second template hex encoded>
```

You can have as many `template_x` columns as you like. The output CSV would look like:

```csv
id_1, id_2, score_1, score_2
<first subject's id>, <second subject's id>, <score from comparing template_1>, <score from comparing template_2>
```

I ended up reworking a bit of the code from Kafis, but really only at a structural level and with some tweaks to how the unique combinations are created (to make the structure we deal with "flatter"). I've kept the `ExecutorService` implementation as-is for now.

**Comment by tobiasmcnulty** [comment]
@seadowg Nice work! We'll take a look and try to test it sometime soon.

CC @copelco

**Comment by seadowg** [comment]
> Yes, Kafis handled 26000+ rows in under 2 hours.

Interesting! I'll need to have a look and see what differences could be causing the problems. @tobiasmcnulty is this a priority for you over Aratek support?

**Comment by tobiasmcnulty** [comment]
@seadowg Thanks for checking. Both are long-term goals for us, so no strong preference either way.

**Comment by seadowg** [comment]
@HusnaHariz also: do you have a test script you can share?

**Comment by HusnaHariz** [comment]
Hi @seadowg.

I have tested the changes, and they were effective in resolving the Java heap space error issue. This was verified with over 26,000 records. 

However, the processing time is still longer than that of Kafis. 

Below is a summary of the observed performance based on input CSV size:

1. 100 records – 1min 51s
2. 500 records – 43min 49s
3. 1000 records – 2h 50min 56s
4. 26,000+ records – Stopped it after 3 hours

**Comment by seadowg** [comment]
@HusnaHariz got it! Could you copy the commands you're running for each for me? Just want to check the args you're using.

**Comment by HusnaHariz** [comment]
Here are the commands:

```
docker run --rm \
  -v "$(pwd)":/data \
  ghcr.io/caktus/odk_biometrics:sha-d9f751f \
  pmatch -i /data/keppel_input.csv -o /data/keppel_outpu.csv -t 40 -p 16
```

Also tested it out before with a maximum heap size env when I was getting the heap space error. 
```
docker run --rm \
  -v "$(pwd)":/data \
  -e JAVA_TOOL_OPTIONS="-Xmx6g" \
  ghcr.io/caktus/odk_biometrics:sha-d9f751f \
  pmatch -i /data/keppel_input.csv -o /data/keppel_output.csv -t 40 -p 16
```

**Comment by tobiasmcnulty** [comment]
@seadowg To clarify one point, the speedup achieved by kafis was largely due to importing the templates once up front, instead of for every match operation:

https://github.com/caktus/kafis/blob/main/src/main/kotlin/com/smartelect/App.kt#L97-L98

This is noted in the [SourceAFIS documentation](https://sourceafis.machinezoo.com/java):

> It is important to call FingerprintMatcher(FingerprintTemplate) constructor only once, because it is an expensive operation. 


Hope that helps!

**Comment by seadowg** [comment]
> Hope that helps!

Massively! That would have taken some profiling (or a lucky guess) to figure out. I can experiment with a solution that get us the same performance profile, while keeping things testable.

**Comment by seadowg** [comment]
@HusnaHariz I think I've managed to get to a place where the speed-up @tobiasmcnulty should be implemented. Let me know what times you're seeing with your data set! I think there are potentially some further improvements that can be made, but let's check we're at parity before going any further.

**Comment by HusnaHariz** [comment]
@seadowg 

There was a significant performance improvement for records under 1000 rows.

500 rows   ~ 1 min 40 secs (down from ~ 44 mins)
1000 rows  ~ 6 mins 39 secs (down from ~ 3 hours)

However, running the changes against 5000 records resulted in a Java Heap Space Error.

**Comment by seadowg** [comment]
@HusnaHariz I can have a look at that soon.

**Comment by seadowg** [comment]
@HusnaHariz one thing to clarify: going back to the Kafis code, I'm realizing that I think Keppel might output differently which could be the reason we're seeing memory fill up. As far as I can tell, Kafis just outputs how many matches there were and how long matching took. Is that correct? If not, I must have access to the wrong code (or just be missing something).

Keppel writes the matches out to a file and I think the actual result collection could be responsible for the heap space errors. I'm going to investigate this further, but just wanted to check that the difference in behaviour is clear.

**Comment by seadowg** [comment]
@HusnaHariz the latest version should help with the heap space errors.

**TLDR**: it should (hopefully) be faster now, and you can tune performance with both parallelism/thread count (using `-p`) and the window size (`-p`). Higher values of each should lead to faster times, but higher memory usage. The default values are 2 and 100 respectively, but I'd imagine you can start experimenting with something like 16 (the thread count it looks like you've been using) and a 1000 - matching prints with 16 threads in batches of a 1000.

> As far as I can tell, Kafis just outputs how many matches there were and how long matching took. Is that correct?

~~It initially seemed to me that Kafis was able to avoid the memory problems we've seen due to the fact that it just deals in integer counts - the values sitting around in memory waiting to be consumed are pretty small. Reading [the code](https://github.com/caktus/kafis/blob/main/src/main/kotlin/com/smartelect/App.kt) again however, I'm now pretty convinced that it's also because Kafis doesn't actually perform matching in parallel:~~

```
combinations.map { (subject, candidates) ->
    workerPool.submit<Long> {
        matchSingleSubject(subject, candidates)
    }
}.sumOf { future ->
    future.get() ?: 0L
} // sum of candidate counts for all subjects
```

~~Here, because `combinations` is a `Sequence` the iteration for `sumOf` will happen one item at a time and the "upstream" operations on the sequence will also be evaluated one item at a time (see `Sequence` docs [here](https://kotlinlang.org/docs/sequences.html)). This means that the `submit` call doesn't happen until we're evaluating an item for `sumOf` and then we immediately block using `get` - the next `submit` call waits for `get` to complete. So yes, each individual match happens asynchronously to the iteration, but the iteration blocks on that async work meaning this is essentially just one linear sweep through the pairs. I'm confused by this as I'd swear I'd seen benchmarks where different thread counts gave you different results, but maybe I've invented this. Or, maybe this code is not the code that was used for benchmarking?~~

EDIT: Ignore the above. I've realized `combinations` is a `List` in the above code which explains how it runs faster than sequentially - the `map` calls will be starting operations. I am surprised Kafis doesn't run into heap space errors due to having all prints in memory, 

I'd fallen for the same trap in my (most recent) implementation, but because my implementation accumulated results (pairs, match scores etc) it still used up memory very quickly. When trying to optimize this, I noticed the flaw. I've ended up with an approach that uses a fixed memory "window": a set amount of print pairs (customized by `-w`) are processed on a set number of threads (customized by `-p`) iteratively (basically a "Cursor"). This means we still find matches in parallel, but we avoid flooding memory with results waiting to be written out to the output CSV.
